### ISSUE TITLE
Async driver support for mssql dialect

### ISSUE THREAD

**Issue by basiltt** [sqlalchemy/sqlalchemy#6521]
Unable to use the aiodbc driver with sqlalchemy. I'm trying to create an async API using FastAPI, But unfortunately in my organization we only have MSSQL servers how can I achieve async ORM for MSSQL DB ?

**Versions.**
 - OS: Windows 10
 - Python: 3.8
 - SQLAlchemy: 1.4.15
 - Database: MSSQL
 - DBAPI: aioodbc

**Comment by CaselIT** [comment]
Hi,

At the moment SQLAchemy does not yet support any driver for mssql (and oracle) using the async access patter. Currently supported dbs using asyncio are Postgresql, Mysql/mariadb and sqlite.

I'll leave this open as a feature request. We currently don't have an eta for the this.
Looking at the library you mentioned, it seems to fairly similar to pyodbc so that should simplify thinks somewhat https://github.com/aio-libs/aioodbc#basic-example

**Comment by gordthompson** [comment]
In working up the example code for [this comment](https://github.com/mkleehammer/pyodbc/issues/901#issuecomment-869225912) I discovered that with aioodbc, `fast_executemany = True` does not cause an error but it also seems to have no effect. Regardless of that setting an aioodbc `.executemany()` appears in SQL Profiler as

```
exec sp_prepexec @p1 output,N'@P1 nvarchar(8)',N'INSERT INTO aioodbc_test (txt) VALUES (@P1)',N'Alfa'
exec sp_prepexec @p1 output,N'@P1 nvarchar(10)',N'INSERT INTO aioodbc_test (txt) VALUES (@P1)',N'Bravo'
```

which is how a pyodbc `.executemany()` appears when `fast_executemany = False`. For pyodbc with `fast_executemany = True` an `.executemany()` appears in SQL Profiler as

```
exec sp_prepare @p1 output,N'@P1 nvarchar(50)',N'INSERT INTO aioodbc_test (txt) VALUES (@P1)',1
exec sp_execute 1,N'Alfa'
exec sp_execute 1,N'Bravo'
```

Depending on the usage case, that could limit the appeal of an `mssql+aioodbc` implementation.

**Comment by gordthompson** [comment]
Proof-of-concept using FastAPI's "SQL database" tutorial with `mssql+pyodbc://…`:

https://github.com/gordthompson/fastapi-tutorial-mssql

**Comment by yormanb** [comment]
amigo una consulta.. baje todo tu codigo cambie las variables para conectar a mi sql.. pero me da este error..
![image](https://user-images.githubusercontent.com/41123797/153972745-d4584502-19b7-4b91-b56f-c92920acd86f.png)

![image](https://user-images.githubusercontent.com/41123797/153972685-164aac7d-4491-4d6d-9fd3-62085c9e15eb.png)

**Comment by tomsoliver** [comment]
> Note also that the [aioodbc README](https://github.com/aio-libs/aioodbc/blob/master/README.rst) says that
> 
> > Internally _aioodbc_ employs threads to avoid blocking the event loop
> 
> and FastAPI's documentation on [Concurrency and async / await](https://fastapi.tiangolo.com/async/#path-operation-functions) says
> 
> > When you declare a _path operation function_ with normal `def` instead of `async def`, it is run in an external threadpool that is then awaited, instead of being called directly (as it would block the server).
> 
> Therefore for FastAPI there is no real point in adding explicit support for aioodbc to the mssql dialect because it would just do the same thing that FastAPI does already when `def` instead of `async def` is used to declare a path operation function.

While it is true that FastAPI runs sync or async path operations in an external threadpool, it does not mean that any code run by those path operations is run asynchronously on seperate threads. For example, a request to `GET /details` may make a request to a SQL database and an API. If they are both syncronous clients, then the call to the API will have to wait for the SQL call to finish. If we want those two tasks can be done asynchronously one needs to utilise an async SQL connection and async HTTP connection.

**Comment by tomsoliver** [comment]
@gordthompson - Yes I do agree, I had a look into it this afternoon and it seems like a lot of effort to support a technology that we are slowly replacing. The benefits of async drivers only really manifest for high performance systems which generally don't use MSSQL.

**Comment by henkepa** [comment]
> The benefits of async drivers only really manifest for high performance systems which generally don't use MSSQL

LOL. I bet more performance systems use MSSQL than PYTHON

**Comment by FranzForstmayr** [comment]
It seems like aioodbc is alive (again), they made a release a few months ago.
https://github.com/aio-libs/aioodbc/releases/tag/v0.4.0

Any relevant updates for this issue?

**Comment by CaselIT** [comment]
> It seems like aioodbc is alive (again), they made a release a few months ago.

Good to hear 

> Any relevant updates for this issue?

Not really, no.

**Comment by FranzForstmayr** [comment]
Can you give some hints what would bei needed to enable support?
Do you see any obvious blockers?

**Comment by CaselIT** [comment]
> Can you give some hints what would bei needed to enable support?

not sure I understand the question. if you ask what's to be done, well someone would have to write support for it. The most similar dialect is probably aiosqlite.

Not 100% sure if the best place for such a dialect would be sqlalchemy or a 3rd party library.

> Do you see any obvious blockers?

well aioodbc is just a thread pool, so it's not really any better that running the sqlalchemy statements in a thread pool.

Other than that pyodbc is not great as a driver. I guess aiodbc can't be much worse though, since it's mostly just pyodbc (hopefully).

**Comment by chen2073** [comment]
> > Can you give some hints what would bei needed to enable support?
> 
> not sure I understand the question. if you ask what's to be done, well someone would have to write support for it. The most similar dialect is probably aiosqlite.
> 
> Not 100% sure if the best place for such a dialect would be sqlalchemy or a 3rd party library.
> 
> > Do you see any obvious blockers?
> 
> well aioodbc is just a thread pool, so it's not really any better that running the sqlalchemy statements in a thread pool.
> 
> Other than that pyodbc is not great as a driver. I guess aiodbc can't be much worse though, since it's mostly just pyodbc (hopefully).

I would like to help adding aioodbc to sqlalchemy's dialect support, how do I get started?😊

**Comment by dbvdb** [comment]
I want to use async sqlalchemy with aioodbc, any updates?

**Comment by CaselIT** [comment]
You will see them here when there are

**Comment by dbvdb** [comment]
Looking forward to see this feature in the next release 😊

**Comment by VitalyKrenel** [comment]
Hey guys! +1 on this request so you could see that there's open request from the community to have async driver with MS SQL

**Comment by sqla-tester** [comment]
**Mike Bayer** has proposed a fix for this issue in the **main** branch:

**add aioodbc support for SQL Server** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4892

**Comment by EmilioGarCan** [comment]
For those encountering this issue - seems like the aioodbc driver for MS SQL that was created has an issue with async commits, and might not fit your use case (this was the case for me). Commits using this driver cause blocks in my API even when they're properly awaited - not sure if it's specific to Azure SQL Server.

I migrated to Postgres along with its respective async driver and the issue was fixed without any further changes to my async workflow or API configuration. Keep an eye out for this if you plan on using the mssql+aioodbc driver.
